### PR TITLE
Langjian/fused batch norm grad

### PR DESCRIPTION
**Pull request by jianyinglang**

### CODE REVIEW

File: src/ngraph_builder.cc [hunk 1]
```diff
@@ -1348,7 +1334,104 @@ tf::Status Builder::TranslateGraph(const std::vector<tf::TensorShape>& inputs,
...
+      float tf_epsilon;
+      if (tf::GetNodeAttr(op->attrs(), "epsilon", &tf_epsilon) !=
+          tf::Status::OK()) {
+        NGRAPH_VLOG(3) << "epsilon attribute not present, setting to zero";
+        tf_epsilon = 0.0001; 
+      }
+
+      NGRAPH_VLOG(3) << "epsilon: " << tf_epsilon;
+
+      // ToDo: We are temporarily supplying a fake value for beta here
```
Comment (aprocter) [inline src/ngraph_builder.cc]: `TODO` should be in all caps

File: src/ngraph_builder.cc [hunk 2]
```diff
@@ -1348,7 +1334,104 @@ tf::Status Builder::TranslateGraph(const std::vector<tf::TensorShape>& inputs,
...
+      // (all zero, same shape/et as scale/gamma), because Tensorflow does not give beta to us.
+      // This should work because nGraph should not actually use beta. The nGraph
+      // op may change to discard this parameter. Update this when nGraph does.
+      shared_ptr<ng::Node> ng_beta =
+          std::make_shared<ngraph::op::Constant>(ng_scale->get_element_type(),
+                           ng_scale->get_shape(),
+                           std::vector<std::string>{ng::shape_size(ng_scale->get_shape()),"0"});
+
+      if (is_nhwc) {
```
Comment (aprocter) [inline src/ngraph_builder.cc]: FYI, @megawidget has a nice refactor in progress for this sort of thing at #94. Not merged yet.

File: test/tf_exec.cpp [hunk 3]
```diff
@@ -280,6 +278,68 @@ TEST(tf_exec, BatchMatMul_2D) {
...
+  attrs.epsilon_ = 0.0001f; 
+  attrs.data_format_ = "NHWC"; 
+
+  std::vector<tf::Tensor> outputs_x;
+  std::vector<tf::Tensor> outputs_scale;
+  std::vector<tf::Tensor> outputs_offset;
+  tf::ClientSession session(dev_scope);
+  auto R = tf::ops::FusedBatchNormGrad(dev_scope.WithOpName("R"), tf_delta, tf_input, 
+                    tf_gamma, tf_mean, tf_variance, attrs);
+  TF_CHECK_OK(session.Run({R.x_backprop}, &outputs_x));
```
Comment (aprocter) [inline test/tf_exec.cpp]: I think there must be a way to do all three of these in a single `Run`.

File: src/ngraph_builder.cc (hunk 2 above)
Comment (megawidget) [reply src/ngraph_builder.cc]: it would be nice if mine went in first, as much as i like merging conflicts :D

**Comment by avijit-nervana** [comment]
@jianyinglang Please respond to @aprocter's comments.

File: src/ngraph_builder.cc (hunk 1 above)
Comment (jianyinglang) [reply src/ngraph_builder.cc]: Changed

File: test/tf_exec.cpp (hunk 3 above)
Comment (jianyinglang) [reply test/tf_exec.cpp]: Changed the three runs to a single `Run`.

File: src/ngraph_builder.cc (hunk 2 above)
Comment (jianyinglang) [reply src/ngraph_builder.cc]: Should I do a cherry-pick from PR #94 or wait it be merged into master?

Comment (avijit-nervana) [reply src/ngraph_builder.cc]: Now #94 is merged into master - so you may want to merge master to your branch and use the helper functions.

**Comment by avijit-nervana** [comment]
@jianyinglang thanks for the changes. Now that Igor's #94 is merged, please update your PR using those functions. Thanks.

File: src/ngraph_builder.cc (hunk 2 above)
Comment (jianyinglang) [reply src/ngraph_builder.cc]: Made the change for `nhwc` case using helper function.

File: src/ngraph_builder.cc [hunk 4]
```diff
@@ -1192,6 +1192,86 @@ tf::Status Builder::TranslateGraph(const std::vector<tf::TensorShape>& inputs,
...
+      }
+
+      bool is_nhwc = (tf_data_format == "NHWC");
+      
+      NGRAPH_VLOG(3) << "data_format: " << tf_data_format;
+
+      float tf_epsilon;
+      if (tf::GetNodeAttr(op->attrs(), "epsilon", &tf_epsilon) !=
+          tf::Status::OK()) {
+        NGRAPH_VLOG(3) << "epsilon attribute not present, setting to zero";
```
Comment (aprocter) [inline src/ngraph_builder.cc]: 0.0001

File: test/tf_exec.cpp [hunk 5]
```diff
@@ -54,7 +54,6 @@ TEST(tf_exec, hello_world) {
   LOG(INFO) << outputs[0].matrix<float>();
 }
 
-#if !defined(NGRAPH_EMBEDDED_IN_TENSORFLOW)
```
Comment (aprocter) [inline test/tf_exec.cpp]: Why this change?

File: src/ngraph_builder.cc (hunk 4 above)
Comment (jianyinglang) [reply src/ngraph_builder.cc]: Changed

File: test/tf_exec.cpp (hunk 5 above)
Comment (jianyinglang) [reply test/tf_exec.cpp]: Not sure how this happened. Changed it back.

**Comment by jianyinglang** [comment]
@avijit-nervana I have merge it to the master and package this `FusedBatchNormGrad` op as a separate function and add the entry in the dispatch table.
